### PR TITLE
[CS] Remove `protected_to_private` rule

### PR DESCRIPTION
**Pull request by SpacePossum**
| Q             | A
| ------------- | ---
| Branch?       | 3.1
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | n/a
| License       | MIT

Following:
https://github.com/symfony/symfony/pull/24123#issuecomment-328031180

I picked 3.1 as target as the rule discussed; 

| branch | hits
| --------- | ---
| 2.7 | 0
| 2.8 | 0
| 3.0 | 0
| 3.1 | 1
| master |  1 hit (same file as on 3.1)

### CODE REVIEW

File: .php_cs.dist [hunk 1]
```diff
@@ -37,5 +37,7 @@ return PhpCsFixer\Config::create()
             ->notPath('Symfony/Bundle/FrameworkBundle/Tests/Templating/Helper/Resources/Custom/_name_entry_label.html.php')
             // explicit heredoc test
             ->notPath('Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Resources/views/translation.html.php')
+            // explict use `protected` attribute in `final` class
```
Comment (keradus) [inline .php_cs.dist]: it's hard without reasoning :( yet, we don't know reasoning

Comment (stof) [inline .php_cs.dist]: typo: `explicit`

**Review by stof — CHANGES_REQUESTED** [review]

File: .php_cs.dist (hunk 1 above)
Comment (stof) [reply .php_cs.dist]: btw, `explicit` seems weird in this case. `Intended` may be better

**Comment by keradus** [comment]
> The "/tmp/fabbot/m_75504550859b26057bb9fe5.98914794/src" directory does not exist.

nice crash @fabpot

**Comment by keradus** [comment]
damn, all CIs are crashing...

**Comment by chalasr** [comment]
3.1 is EOMed, this should be rebased on 3.3.  That will trigger a new fabbot's review btw (other CIs might be broken, we are able to know if it's related or not).

**Comment by SpacePossum** [comment]
I'm not sure I understand I should do (since I don't know what 'EOMed' means) 
Shouldn't 3.1 be the target as it is the lowest branch effected?

**Comment by chalasr** [comment]
3.1 is not maintained anymore (EOM => End Of Maintenance), thus the lowest maintained branch affected by this should be 3.3.

**Comment by lyrixx** [comment]
> The "/tmp/fabbot/m_75504550859b26057bb9fe5.98914794/src" directory does not exist.
> nice crash @fabpot

The crash does not come from fabbot. It's because of the PHPCSFixer [config](https://github.com/symfony/symfony/blob/8357f2ae4ab490e4b561317ebfd99cc09a09d885/.php_cs.dist#L16). 

Fabbot only write new/updated files on the disk. In this case, there are no file updated in the src/ folder. So the finder crashes.

We should add a `if is_dir` to prevent this.

**Comment by nicolas-grekas** [comment]
@jvasseur any comment?

**Comment by jvasseur** [comment]
This means we will stop getting any CS fix for this file. I think we should stop `php-cs-fixer` from doing the `protected` -> `private` change for the whole codebase instead, since it doesn't bring a lot of value (we already push for using private everywhere).

BTW since this fixer can change code behavior (event if it's a bit of an edge case) it should be marked as risky.

**Comment by keradus** [comment]
any fixer could be marked as risky if you violate php or have test that compare how code "looks like" (like here, dumper of class - `null` vs `NULL`).
I rather suggest you to stop violating PHP OOP in CacheItem ;)
or, continue violating it if you really need to, yet change way of accessing variable from protected to private - it's still possible, just use different key prefix.

**Comment by nicolas-grekas** [comment]
@SpacePossum is there a way to achieve what @jvasseur says? ie exclude that protected->private transformation?

**Comment by nicolas-grekas** [comment]
@keradus I don't know who's that "you" you're referring to, but that's OSS, so if you want to propose a change, you can submit PRs and trigger a discussion on actual code, not only talk in the abstract. I'd be happy to explain on that PR why I think the current code has merits that surpasses the principles you defend. Changing to private and accessing with a different key is also possible. I'll let you do the PR.
Until this is done, I agree with @jvasseur, the CS fixer is just wrong here, we don't follow those rules, the code proves it.

**Comment by SpacePossum** [comment]
>This means we will stop getting any CS fix for this file. 

Yes, this is sadly the side effect of the proposed change. I say "sadly" because I would like a way of configuring rules for different directories. However, reality is I didn't find the time to make it.

>I think we should stop php-cs-fixer from doing the protected -> private change for the whole codebase instead, since it doesn't bring a lot of value (we already push for using private everywhere).

I think to would be a loss for SF. It is not about the good code you got not now, it is more to prevent it from happening in the future. Good reviews help, but having a tool to do it for you saves time for reviewers pointing out stuff and focus on the real value of a PR.

> BTW since this fixer can change code behavior (event if it's a bit of an edge case) it should be marked as risky.

I didn't look into the implementation of the fixer for a while so I cannot say anything useful about it now. Therefore I would like to advice you to open an issue about it on the fixer repo itself. I tend to forget requests like these and an issue there might prevent that, but feel free to ignore my suggestion.
One note though, since the current configuration allows [risky](https://github.com/symfony/symfony/blob/3.4/.php_cs.dist#L13) rules moving the fixer to the SF risky set will still cause fabbot.io to report it.

>any fixer could be marked as risky if you violate php or have test that compare how code "looks like" (like here, dumper of class - null vs NULL).

Yes, anyone relying on the source itself through `dumper`, reflection, tokenizing, file size, etc. will find that all fixers are risky.
We consider `risky` possibly changing code behavior, however we try to support other systems relying on PHPDoc (Doctrine for example) as well and not break those (even if it wouldn't break PHP code execution itself) (but than again, you know way better than me :) ). 

> I rather suggest you to stop violating PHP OOP in CacheItem ;)
or, continue violating it if you really need to, yet change way of accessing variable from protected to private - it's still possible, just use different key prefix.

Using the different key-prefix sounds like a good idea to me.

>@SpacePossum is there a way to achieve what @jvasseur says? ie exclude that protected->private transformation?

You can remove the rule as a whole or remove all rules from being applied to a file (this PR).
Sadly no middle ground like `apply only this set to these files`.

>@keradus I don't know who's that "you" you're referring to, but that's OSS, so if you want to propose a change, you can submit PRs and trigger a discussion on actual code, not only talk in the abstract. I'd be happy to explain on that PR why I think the current code has merits that surpass the principles you defend. Changing to private and accessing with a different key is also possible. I'll let you do the PR.

I think a dedicated issue about this would work as well? I would be interested as for me this has been an TIL day about PHP while I thought I knew a bit about it :)

>Until this is done, I agree with @jvasseur, the CS fixer is just wrong here, we don't follow those rules, the code proves it.

Not following a rule doesn't make the fixer wrong, it means it is configured in a way that doesn't suit you.
The difference I try to point out is that we (PHP CS Fixer community) do try to provide a good tool that is useful for SF and others. We try to keep the amount of noise down for SF based on previous PR's and such. We might get it wrong sometimes, but we also try to follow up on that and fix the mistakes.

This PR can also wait until the talk about the file has been finished. It can also be closed as not relevant. For me it was just an attempt to reduce the noise caused by the fixer on the file, if it causes even more noise; than the PR should be stopped ;)

**Comment by chalasr** [comment]
Yet, PHP allows defining protected members in a final class. About "why we rely on that", this is a framework after all, we have needs that final applications don't have. 
Same principle applies for `func_get_args()`, reflection, closure binding and other stuff that you might not need (or want) in your final applications. 

:+1: for disabling this rule in the core rather than excluding this file from the inspection. This file not a hack, excluding it would be.

**Comment by nicolas-grekas** [comment]
> Not following a rule doesn't make the fixer wrong, it means it is configured in a way that doesn't suit you

100% what I meant, php-cs-fixer is a fantastic tool indeed, a big thank for making it alive really! I only meant configuration was not aligned to actual practice. Can you please drop a PR disabling that rule (and enabling yoda_style & co as discussed in the previous PR?)

**Comment by keradus** [comment]
> configuration was not aligned to actual practice

debatable, whole repo except one file is following it ;)

**Comment by lyrixx** [comment]
:+1:

**Comment by SpacePossum** [comment]
updated,
lemme know if I should remove the commit from @lyrixx (I picked it to see `fabbot.io` running here)

**Comment by nicolas-grekas** [comment]
@SpacePossum thanks, I just merged @lyrixx' commit up to 3.3 so you can rebase and get rid of it here.

**Comment by SpacePossum** [comment]
Thanks so much :+1: 
commit has been removed by rebasing

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @SpacePossum.

**Pull request merged**